File: backend/src/application/use-cases/shopping-session/DeleteShoppingSessionUseCase.ts
import type { DeleteShoppingSessionUseCasePort } from "@application/ports/incoming/DeleteShoppingSessionUseCasePort";
import type { ShoppingSessionRepository } from "@application/ports/outgoing/ShoppingSessionRepository";
import type { LoggerPort } from "@application/ports/outgoing/LoggerPort";
import type {
	DeleteShoppingSessionInput,
	DeleteShoppingSessionResult,
} from "./contracts";

export class DeleteShoppingSessionUseCase
	implements DeleteShoppingSessionUseCasePort
{
	constructor(
		private readonly shoppingSessionRepository: ShoppingSessionRepository,
		private readonly logger: LoggerPort,
	) {}

	async execute(
		input: DeleteShoppingSessionInput,
	): Promise<DeleteShoppingSessionResult> {
		if (input.sessionId.trim().length === 0) {
			throw new Error("sessionId is required.");
		}

		if (input.userId.trim().length === 0) {
			throw new Error("userId is required.");
		}

		this.logger.info(
			`[DeleteShoppingSessionUseCase] execute - userId: "${input.userId}", sessionId: "${input.sessionId}"`,
		);

		const deleted = await this.shoppingSessionRepository.deleteByIdForUser(
			input.sessionId,
			input.userId,
		);

		return { deleted };
	}
}

File: backend-dotnet/src/api/Application/UseCases/ShoppingSession/DeleteShoppingSessionUseCase.cs
using Microsoft.Extensions.Logging;
using SharpMonads.Core;
using SupermarketPlanner.Api.Application.Ports.Driven;
using SupermarketPlanner.Api.Application.Ports.Driving;
using SupermarketPlanner.Api.Domain;

namespace SupermarketPlanner.Api.Application.UseCases.ShoppingSession;

public sealed class DeleteShoppingSessionUseCase(
    IShoppingSessionRepository shoppingSessionRepository,
    ILogger<DeleteShoppingSessionUseCase> logger) : IDeleteShoppingSessionUseCase
{
    public Task<Either<DomainError, DeleteShoppingSessionResult>> Invoke(
        DeleteShoppingSessionInput input,
        CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(input.SessionId))
            return Task.FromResult(Either<DomainError, DeleteShoppingSessionResult>.FromLeft(
                DomainError.Validation(nameof(input.SessionId), "sessionId is required.")));

        if (string.IsNullOrWhiteSpace(input.UserId))
            return Task.FromResult(Either<DomainError, DeleteShoppingSessionResult>.FromLeft(
                DomainError.Validation(nameof(input.UserId), "userId is required.")));

        logger.LogInformation(
            "[DeleteShoppingSessionUseCase] Invoke - userId: \"{UserId}\", sessionId: \"{SessionId}\"",
            input.UserId, input.SessionId);

        return shoppingSessionRepository.DeleteByIdForUser(input.SessionId, input.UserId, cancellationToken)
            .MapAsync(deleted => new DeleteShoppingSessionResult(Deleted: deleted));
    }
}
